### PR TITLE
Adds add2app Flutter module and an iOS project to run it.

### PR DESCRIPTION
**Pull request by RedBrogdon**

### CODE REVIEW

**Comment by xster** [comment]
also cc @jmagman

File: experimental/add2app/example_module/pubspec.yaml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+version: 1.0.0
+
+environment:
+  sdk: ">=2.4.0 <3.0.0"
+
+dependencies:
+  flutter:
+    sdk: flutter
+
```
Comment (jmagman) [inline experimental/add2app/example_module/pubspec.yaml]: It would be nice to add at least one plugin to show how plugin registration works in SimpleIOSExample.

Comment (RedBrogdon) [reply experimental/add2app/example_module/pubspec.yaml]: +1.

I'm starting off with something basic, but at some point I'll go round to everyone working on add2app to make sure I'm demonstrating everything you feel should be covered.

**Comment by johnpryan** [comment]
I'm seeing an error running `pod install`:

```
pod install
Ignoring eventmachine-1.2.7 because its extensions are not built.  Try: gem pristine eventmachine --version 1.2.7
Ignoring executable-hooks-1.6.0 because its extensions are not built.  Try: gem pristine executable-hooks --version 1.6.0
Ignoring ffi-1.11.1 because its extensions are not built.  Try: gem pristine ffi --version 1.11.1
Ignoring ffi-1.9.25 because its extensions are not built.  Try: gem pristine ffi --version 1.9.25
Ignoring gem-wrappers-1.4.0 because its extensions are not built.  Try: gem pristine gem-wrappers --version 1.4.0
Ignoring http_parser.rb-0.6.0 because its extensions are not built.  Try: gem pristine http_parser.rb --version 0.6.0
Ignoring nokogiri-1.8.5 because its extensions are not built.  Try: gem pristine nokogiri --version 1.8.5
Ignoring nokogiri-1.8.4 because its extensions are not built.  Try: gem pristine nokogiri --version 1.8.4
Ignoring sassc-2.0.0 because its extensions are not built.  Try: gem pristine sassc --version 2.0.0
Ignoring sassc-1.12.1 because its extensions are not built.  Try: gem pristine sassc --version 1.12.1

[!] Invalid `Podfile` file: cannot load such file -- ../example_module/.ios/Flutter/podhelper.rb.

 #  from /Users/ryjohn/code/github/flutter/samples/experimental/add2app/SimpleIOSExample/Podfile:6
 #  -------------------------------------------
 #  flutter_application_path = '../example_module'
 >  load File.join(flutter_application_path, '.ios', 'Flutter', 'podhelper.rb')
 #  
 #  -------------------------------------------
```

**Comment by RedBrogdon** [comment]
@johnpryan I ran into the same issue while testing a clone. There's a pod-helper script that's created by `flutter create` and `flutter build`, but not meant to be checked in. As a result, if you do a `pod install` first thing after cloning, the script isn't there and CocoaPods bombs.

I've added a line to the podfile that will run `flutter create` on your behalf automatically to fix this. Can you pull the latest changes and try again?

File: experimental/add2app/SimpleIOSExample/Podfile [hunk 2]
```diff
@@ -0,0 +1,26 @@
+# Uncomment the next line to define a global platform for your project
+# platform :ios, '9.0'
+
+# Added these two lines for Flutter:
+flutter_application_path = '../example_module'
+system("pushd #{flutter_application_path} && flutter create . && popd")
```
Comment (jmagman) [inline experimental/add2app/SimpleIOSExample/Podfile]: Can you use backticks here instead?  Also, handle if flutter_application_path has spaces in it.
```
`pushd "#{flutter_application_path}" && flutter create . && popd`
```

Comment (jmagman) [reply experimental/add2app/SimpleIOSExample/Podfile]: (In this case it doesn't have spaces, but as an example that could be pasted somewhere else, probably good to handle it).

**Comment by johnpryan** [comment]
@RedBrogdon I'm still running into a missing "podhelper.rb" file.

Here's my steps:

```
git@github.com:flutter/samples.git
cd samples
git checkout -b RedBrogdon-add2app master
git pull git://github.com/RedBrogdon/samples.git add2app
cd experimental/add2app/SimpleIOSExample/
pod install
```

Here's the output:

```
Ignoring eventmachine-1.2.7 because its extensions are not built.  Try: gem pristine eventmachine --version 1.2.7
Ignoring executable-hooks-1.6.0 because its extensions are not built.  Try: gem pristine executable-hooks --version 1.6.0
Ignoring ffi-1.11.1 because its extensions are not built.  Try: gem pristine ffi --version 1.11.1
Ignoring ffi-1.9.25 because its extensions are not built.  Try: gem pristine ffi --version 1.9.25
Ignoring gem-wrappers-1.4.0 because its extensions are not built.  Try: gem pristine gem-wrappers --version 1.4.0
Ignoring http_parser.rb-0.6.0 because its extensions are not built.  Try: gem pristine http_parser.rb --version 0.6.0
Ignoring nokogiri-1.8.5 because its extensions are not built.  Try: gem pristine nokogiri --version 1.8.5
Ignoring nokogiri-1.8.4 because its extensions are not built.  Try: gem pristine nokogiri --version 1.8.4
Ignoring sassc-2.0.0 because its extensions are not built.  Try: gem pristine sassc --version 2.0.0
Ignoring sassc-1.12.1 because its extensions are not built.  Try: gem pristine sassc --version 1.12.1
~/code/github/flutter/samples/experimental/add2app/example_module ~/code/github/flutter/samples/experimental/add2app/SimpleIOSExample
Recreating project ....
  example_module.iml (created)
  example_module_android.iml (created)
  .idea/libraries/Flutter_for_Android.xml (created)
  .idea/libraries/Dart_SDK.xml (created)
  .idea/modules.xml (created)
  .idea/workspace.xml (created)
Because every version of flutter_test from sdk depends on pedantic 1.7.0 and example_module depends on pedantic ^1.8.0+1, flutter_test from sdk is forbidden.
So, because example_module depends on flutter_test any from sdk, version solving failed.
Running "flutter pub get" in example_module...                          
pub get failed (1)

[!] Invalid `Podfile` file: cannot load such file -- ../example_module/.ios/Flutter/podhelper.rb.

 #  from /Users/ryjohn/code/github/flutter/samples/experimental/add2app/SimpleIOSExample/Podfile:7
 #  -------------------------------------------
 #  system("pushd #{flutter_application_path} && flutter create . && popd")
 >  load File.join(flutter_application_path, '.ios', 'Flutter', 'podhelper.rb')
 #  
 #  -------------------------------------------
```

**Comment by jmagman** [comment]
> @RedBrogdon I'm still running into a missing "podhelper.rb" error.

Just to confirm, you're on the master channel, right?

File: experimental/add2app/README.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+in the `add2app` folder.
+
+### `SimpleIOSExample`
+
+An iOS application that imports `example_module` and uses it to display
+a `FlutterViewController` containing widgets from the module.
+
+## Questions/issues
+
+If you have a general question about JSON serialization in Flutter, the
```
Comment (xster) [inline experimental/add2app/README.md]: Is this related?

File: experimental/add2app/SimpleIOSExample/Podfile (hunk 2 above)
Comment (xster) [inline experimental/add2app/SimpleIOSExample/Podfile]: This line is surprising. What's the intent here?

File: experimental/add2app/SimpleIOSExample/Podfile [hunk 4]
```diff
@@ -0,0 +1,26 @@
+# Uncomment the next line to define a global platform for your project
+# platform :ios, '9.0'
+
+# Added these two lines for Flutter:
```
Comment (xster) [inline experimental/add2app/SimpleIOSExample/Podfile]: Not sure what this comment meant.

File: experimental/add2app/SimpleIOSExample/Podfile [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+flutter_application_path = '../example_module'
+system("pushd #{flutter_application_path} && flutter create . && popd")
+load File.join(flutter_application_path, '.ios', 'Flutter', 'podhelper.rb')
+
+target 'SimpleIOSExample' do
+  # Comment the next line if you don't want to use dynamic frameworks
```
Comment (xster) [inline experimental/add2app/SimpleIOSExample/Podfile]: Is this coming from a template?
I feel like we don't need this.

File: experimental/add2app/SimpleIOSExample/SimpleIOSExample/ViewController.swift [hunk 6]
```diff
@@ -0,0 +1,17 @@
+// Copyright 2019 The Flutter team. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import UIKit
+import Flutter
+
+class ViewController: UIViewController {
+
+    @IBOutlet weak var launchButton: UIButton!
```
Comment (xster) [inline experimental/add2app/SimpleIOSExample/SimpleIOSExample/ViewController.swift]: I'd just not connect it if you're not gonna use it

File: experimental/add2app/example_module/README.md [hunk 7]
```diff
@@ -0,0 +1,3 @@
+# example_module
+
+A simple module used to demo Flutter's Add2App features
```
Comment (xster) [inline experimental/add2app/example_module/README.md]: A simple module to be embedded in other host apps, used to demo Flutter's add-to-app process.

File: experimental/add2app/SimpleIOSExample/Podfile (hunk 2 above)
Comment (jmagman) [reply experimental/add2app/SimpleIOSExample/Podfile]: We have a thread in the Add-to-app chat about this.  The podhelper.rb script is in the .gitignore, so someone who just cloned this repo needs to create it or they can't `pod install`.  I don't really want it to be checked in, but maybe there's a better solution?

**Comment by johnpryan** [comment]
@xster Switching to the master channel fixed it.

Now the XCode 10.3 build is failing:

```
:-1: Unable to load contents of file list: '/Users/ryjohn/code/github/flutter/samples/experimental/add2app/SimpleIOSExample/Pods/Target Support Files/Pods-SimpleIOSExample/Pods-SimpleIOSExample-frameworks-Debug-input-files.xcfilelist' (in target 'SimpleIOSExample')
```

File: experimental/add2app/SimpleIOSExample/Podfile (hunk 2 above)
Comment (xster) [reply experimental/add2app/SimpleIOSExample/Podfile]: Ah ok. Aren't these made by running flutter packages get instead of flutter create?

File: experimental/README.md [hunk 8]
```diff
@@ -25,7 +25,15 @@ flutter channel stable
...
 
-There aren't actually any samples in here just yet. Things are coming, though!
+### [add2app](add2app)
+
+This project shows how to add a Flutter module to existing Android and iOS
+applications. It uses CocoaPods, so before compiling it, you need to do a pod
+install:
+
+```bash
+pod install
```
[the remark this answers is not in the document]
Comment (RedBrogdon) [reply experimental/README.md]: Nope. This line made more sense before I refactored a few things. I've removed it and added better instructions inside the add2app README.

File: experimental/README.md [hunk 9]
```diff
@@ -25,7 +25,15 @@ flutter channel stable
 
 ## Index
 
-There aren't actually any samples in here just yet. Things are coming, though!
+### [add2app](add2app)
```
[the remark this answers is not in the document]
Comment (RedBrogdon) [reply experimental/README.md]: Not a big deal. What about the project path? Would you prefer to see that change from add2app as well?

File: experimental/add2app/README.md (hunk 3 above)
Comment (RedBrogdon) [reply experimental/add2app/README.md]: Nope. That's an oversight. Good catch.

File: experimental/add2app/SimpleIOSExample/Podfile (hunk 4 above)
Comment (RedBrogdon) [reply experimental/add2app/SimpleIOSExample/Podfile]: It makes some sense in the context of the add2app guide, but you're right that it looks odd without it. I've updated it.

File: experimental/add2app/SimpleIOSExample/Podfile (hunk 5 above)
Comment (RedBrogdon) [reply experimental/add2app/SimpleIOSExample/Podfile]: Yes, that was auto-generated by `pod init`. I've removed it.

File: experimental/add2app/SimpleIOSExample/SimpleIOSExample/ViewController.swift (hunk 6 above)
Comment (RedBrogdon) [reply experimental/add2app/SimpleIOSExample/SimpleIOSExample/ViewController.swift]: Fair enough. Removed.

File: experimental/add2app/example_module/README.md (hunk 7 above)
Comment (RedBrogdon) [reply experimental/add2app/example_module/README.md]: Done.

File: experimental/add2app/SimpleIOSExample/Podfile (hunk 2 above)
Comment (RedBrogdon) [reply experimental/add2app/SimpleIOSExample/Podfile]: Both work, apparently. Does `flutter create` also get packages?

Comment (RedBrogdon) [reply experimental/add2app/SimpleIOSExample/Podfile]: Great point. I've updated it.

Comment (RedBrogdon) [reply experimental/add2app/SimpleIOSExample/Podfile]: Strike that, actually. When I try with `flutter packages get`, the script is created, but there's some other part of the module missing. I end up with an error in Xcode that states it can't find `lib/main.dart`.

**Comment by RedBrogdon** [comment]
PTAL.

All comments addressed except for two waiting on replies. The current code builds out of the box on my machine with these steps:

* `cd experimental/add2app/SimpleIOSExample`
* `pod install`
* `open SimpleIOSExample.xcworkspace`
* Run that sucker.

@johnpryan Do you mind checking to see if a fresh copy of this PR builds for you?

**Comment by johnpryan** [comment]
for future reference, `gem install cocoapods` upgraded my cocoapods version to 1.7.5. I was using an older version (something like 1.3)

File: experimental/README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (xster) [reply experimental/README.md]: Yes please. We're trying to avoid 'clever' names and use plain english as much as possible.

File: experimental/add2app/SimpleIOSExample/Podfile (hunk 2 above)
Comment (xster) [reply experimental/add2app/SimpleIOSExample/Podfile]: What are the error outputs? Might be a valid bug.

File: experimental/add2app/SimpleIOSExample/Podfile (hunk 5 above)
Comment (xster) [reply experimental/add2app/SimpleIOSExample/Podfile]: Looks like it's still here :)

File: experimental/add2app/SimpleIOSExample/Podfile (hunk 2 above)
Comment (RedBrogdon) [reply experimental/add2app/SimpleIOSExample/Podfile]: Weird. I can no longer replicate the failure.

**Comment by RedBrogdon** [comment]
@jmagman @xster 

I went ahead and landed this, because a) it's huge, and b) we need to get the basic project structure in the repo, but the app *occasionally* segfaults. Would either of you be interested in taking a look? 😄

**Comment by xster** [comment]
Yes. Please write down all outputs/stacks etc if you ever encounter any.
